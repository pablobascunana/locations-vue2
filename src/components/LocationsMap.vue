<template>
  <div>
    <h2 class="text-center">{{ $t('locations.title') }}</h2>
    <v-container class="pl-0">
      <v-row class="justify-center">
        <v-col xl="10" lg="10" md="8" sm="8" xs="8">
          <gmap-autocomplete
            id="searchLocation"
            class="autocomplete_input"
            @place_changed="setPlace"
            @keyup.enter="addMarker">
          </gmap-autocomplete>
        </v-col>
        <v-btn
          id="addLocation"
          color="primary"
          @click="addMarker"
        >{{ $t('generic.buttons.add' )}}
        </v-btn>
      </v-row>
    </v-container>
    <GmapMap
      ref="mapRef"
      :center="center"
      :zoom="12"
      map-type-id="satellite"
      :options="mapOptions"
      style="width: 100%; height: 500px"
    >
      <GmapMarker
        :key="index"
        v-for="(marker, index) in markers"
        :position="marker.position"
        :clickable="true"
        :draggable="false"
        @click="center=marker.position"
      />
  </GmapMap>
  </div>
</template>

<script>
import Api from '@/api/services';
export default {
  name: 'locationsMap',
  props: {
    markers: { type: Array, default: () => []},
    userUuid: { type: String, required: true }
  },
  data() {
    return {
      center: { lat: 45.508, lng: -73.587 },
      currentPlace: null,
      mapOptions: {
        zoomControl: true,
        mapTypeControl: false,
        scaleControl: false,
        streetViewControl: false,
        rotateControl: false,
        fullscreenControl: true,
        disableDefaultUi: false
      }
    }
  },
  mounted() {
    this.geolocate();
  },
  methods: {
    setPlace(place) {
      this.currentPlace = place;
    },
    addMarker() {
      if (this.currentPlace) {
        const marker = {
          lat: this.currentPlace.geometry.location.lat(),
          lng: this.currentPlace.geometry.location.lng()
        };
        const markersCopy = [...this.markers]
        markersCopy.push({ position: marker, title: this.currentPlace.name });
        this.center = marker;
        this.currentPlace = null;
        this.saveLocation(markersCopy);
      }
    },
    geolocate() {
      navigator.geolocation.getCurrentPosition(position => {
        this.center = {
          lat: position.coords.latitude,
          lng: position.coords.longitude
        };
      });
    },
    async saveLocation(markersCopy) {
      try {
        const body = { markers: markersCopy, userId: this.userUuid };
        let { data } = await Api.saveLocation(this.userUuid, body);
        this.showSuccessSnackBarAndAddMarker(data.markers);
      } catch(error) {
        this.showErrorSnackBar();
      }
    },
    showSuccessSnackBarAndAddMarker(marker) {
      this.EventBus.$emit('showSnackBar', { 
        snackBarMessage: this.$t('locations.success.save'),
      });
      this.markers.push(marker.at(-1));
    },
    showErrorSnackBar() {
      this.EventBus.$emit('showSnackBar', { 
        snackBarMessage: this.$t('locations.error.saveGeneric'),
        snackBarColor: 'error'
      });
    }
  }
}
</script>

<style scoped>
.autocomplete_input {
  width: 100%;
  outline-width: inherit;
  border-bottom: solid 1px;
}
</style>