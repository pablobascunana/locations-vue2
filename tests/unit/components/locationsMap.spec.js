import { createLocalVue, shallowMount } from '@vue/test-utils'

import Vuetify from 'vuetify'

import LocationsMap from '@/components/LocationsMap.vue'

const Plugins = {
    install(Vue) {
      Vue.prototype.EventBus = new Vue();
    }
};

const localVue = createLocalVue();
localVue.use(Plugins);
let vuetify = new Vuetify();

// https://salesforce.stackexchange.com/questions/261538/how-to-mock-navigator-geolocation-in-a-lwc
const mockGeolocation = {
    getCurrentPosition: jest.fn().mockImplementation((success) => 
        Promise.resolve(
            success({
                coords: {
                    latitude: 10,
                    longitude: 10
                }
            })
        )
    )
};

navigator.geolocation = mockGeolocation;

describe('Locations Map', () => {
    let wrapper;
    const mocks = {
        EventBus: {
            $emit: jest.fn(),
        },
        $t: () => {'some specific text in any language'}
    };
    beforeEach(() => {
        wrapper = shallowMount(LocationsMap, {
            localVue,
            vuetify,
            mocks
        });
    });

    it("Should match snapshot", () => {
        expect(wrapper).toMatchSnapshot()    
    });

    it('Renders a vue instance', () => {
        expect(wrapper.isVueInstance()).toBe(true);
    });

    it('Call to set place function', () => {
        const setPlaceFn = jest.spyOn(wrapper.vm, 'setPlace');
        wrapper.vm.setPlace({ geometry: { location: { lat: () => 10, lng: () => 10 } } });
        expect(setPlaceFn).toBeCalled();
        expect(wrapper.vm.currentPlace).not.toBe(null);
    });

    it('Call to add marker function', () => {
        wrapper.vm.currentPlace = { geometry: { location: { lat: () => 10, lng: () => 10 } } };
        const addMarkerFn = jest.spyOn(wrapper.vm, 'addMarker');
        wrapper.vm.addMarker();
        expect(addMarkerFn).toBeCalled();
        expect(wrapper.vm.center).toStrictEqual({ lat: 10, lng: 10 });
        expect(wrapper.vm.currentPlace).toBe(null);
    });

    // TODO review at is not a function problem here
    // it('Call to show success snackBar and add marker function', () => {
    //     const showSuccessSnackBarAndAddMarkerFn = jest.spyOn(wrapper.vm, 'showSuccessSnackBarAndAddMarker');
    //     wrapper.vm.showSuccessSnackBarAndAddMarker(historicalLocationsMock.MARKER);
    //     expect(showSuccessSnackBarAndAddMarkerFn).toBeCalled();
    // });

    it('Call to show error snack bar function', () => {
        const showErrorSnackBarFn = jest.spyOn(wrapper.vm, 'showErrorSnackBar');
        wrapper.vm.showErrorSnackBar();
        expect(showErrorSnackBarFn).toBeCalled();
    });
});