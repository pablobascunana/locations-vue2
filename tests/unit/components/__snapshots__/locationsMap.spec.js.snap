// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Locations Map Should match snapshot 1`] = `
<div>
  <h2 class="text-center"></h2>
  <v-container class="pl-0">
    <v-row class="justify-center">
      <v-col xl="10" lg="10" md="8" sm="8" xs="8">
        <gmap-autocomplete class="autocomplete_input"></gmap-autocomplete>
      </v-col>
      <v-btn color="primary">
      </v-btn>
    </v-row>
  </v-container>
  <gmapmap center="[object Object]" zoom="12" map-type-id="satellite" options="[object Object]" style="width: 100%; height: 500px;"></gmapmap>
</div>
`;

exports[`LocationsMap Should match snapshot 1`] = `
<div>
  <h2 class="text-center"></h2>
  <v-container class="pl-0">
    <v-row class="justify-center">
      <v-col xl="10" lg="10" md="8" sm="8" xs="8">
        <gmap-autocomplete class="autocomplete_input"></gmap-autocomplete>
      </v-col>
      <v-btn color="primary">
      </v-btn>
    </v-row>
  </v-container>
  <gmapmap center="[object Object]" zoom="12" map-type-id="satellite" options="[object Object]" style="width: 100%; height: 500px;"></gmapmap>
</div>
`;

exports[`LocationsMap should match snapshot 1`] = `
<div>
  <h2 class="text-center"></h2>
  <v-container class="pl-0">
    <v-row class="justify-center">
      <v-col xl="10" lg="10" md="8" sm="8" xs="8">
        <gmap-autocomplete class="autocomplete_input"></gmap-autocomplete>
      </v-col>
      <v-btn color="primary">
      </v-btn>
    </v-row>
  </v-container>
  <gmapmap center="[object Object]" zoom="12" map-type-id="satellite" options="[object Object]" style="width: 100%; height: 500px;"></gmapmap>
</div>
`;
